direct-auth-samples/src/main/java/com/okta/spring/example/controllers/LoginController.java

```diff
@@ -0,0 +1,300 @@
+import com.okta.idx.sdk.api.client.IDXClient;
+import com.okta.idx.sdk.api.model.AuthenticationOptions;
+import com.okta.idx.sdk.api.model.AuthenticationStatus;
+import com.okta.idx.sdk.api.model.AuthenticatorType;
+import com.okta.idx.sdk.api.model.AuthenticatorUIOption;
+import com.okta.idx.sdk.api.model.ChangePasswordOptions;
+import com.okta.idx.sdk.api.model.FormValue;
+import com.okta.idx.sdk.api.model.IDXClientContext;
+import com.okta.idx.sdk.api.model.RecoverPasswordOptions;
+import com.okta.idx.sdk.api.model.UserProfile;
+import com.okta.idx.sdk.api.model.VerifyAuthenticatorOptions;
+import com.okta.idx.sdk.api.response.AuthenticationResponse;
+import com.okta.idx.sdk.api.response.NewUserRegistrationResponse;
+import com.okta.idx.sdk.api.wrapper.AuthenticationWrapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.servlet.ModelAndView;
+
+import javax.servlet.http.HttpSession;
+import java.util.List;
+
+@Controller
+public class LoginController {
+
+    private final Logger logger = LoggerFactory.getLogger(LoginController.class);
+
+    @Autowired
+    private IDXClient client;
+
+    @PostMapping("/custom-login")
+    public ModelAndView handleLogin(@RequestParam("username") String username,
+                                    @RequestParam("password") String password,
+                                    HttpSession session) {
+
+        AuthenticationResponse authenticationResponse =
+                (AuthenticationResponse) session.getAttribute("authenticationResponse");
+
+        // render existing auth response if a successful one is already present in session
+        if (authenticationResponse != null &&
+            authenticationResponse.getAuthenticationStatus() == AuthenticationStatus.SUCCESS) {
+            ModelAndView mav = new ModelAndView("home");
+            mav.addObject("authenticationResponse", authenticationResponse);
+            return mav;
+        }
+
+        // trigger authentication
+        authenticationResponse = AuthenticationWrapper.authenticate(client, new AuthenticationOptions(username, password));
+
+        // populate login view with errors
+        if (authenticationResponse.getAuthenticationStatus() != AuthenticationStatus.SUCCESS) {
+            ModelAndView mav = new ModelAndView("login");
+            mav.addObject("errors", authenticationResponse.getErrors());
+            return mav;
+        }
+
+        // success
+        ModelAndView mav = new ModelAndView("home");
+        mav.addObject("authenticationResponse", authenticationResponse);
+
+        // store attributes in session
+        session.setAttribute("user", username);
+        session.setAttribute("authenticationResponse", authenticationResponse);
```

Comment:
Can we set less in the session here? Maybe just the tokens and the idxClientContext?

Reply:
`user` is stored in session because it is used in the view to display `Welcome ${user}`.

`token` is stored inside `authenticationResponse` object. `idxClientContext` is not needed to be stored at this point.

Reply:
updated